iceoryx_posh/source/runtime/service_discovery.cpp

```diff
@@ -21,16 +21,13 @@ namespace iox
 {
 namespace runtime
 {
-ServiceContainer ServiceDiscovery::findService(const cxx::optional<capro::IdString_t>& service,
-                                               const cxx::optional<capro::IdString_t>& instance,
-                                               const cxx::optional<capro::IdString_t>& event) noexcept
+void ServiceDiscovery::update()
 {
-    ServiceContainer searchResult;
-
-    auto lambda = [&](const capro::ServiceDescription& entry) { searchResult.emplace_back(entry); };
-    findService(service, instance, event, lambda);
-
-    return searchResult;
+    // allows us to use update and hence findService concurrently
+    std::lock_guard<std::mutex> lock(m_serviceRegistryMutex);
```

Comment:
YAGNI? Currently `update()` is private and can't be called concurrently by the user. Please either make `update()` public or remove the lock. I would prefer the latter.

Reply:
No, clean design. And stop this YAGNI please.. `update` exists to cleanly separate findService and update. `findService` is calling update. Only the update needs to be concurrently protected (because it writes), not the whole findService (as the remainder only reads). -> Keep critical sections small

I do not want to have bug reports when it is used in multiple threads and the mutex prevents problems here.

Reply:
@MatthiasKillat Is the use case of concurrent `findService` so prominent? Several other parts of the user API is not usable concurrently. In the past we added a comment or doxygen hint for these parts. Why not just add a doxygen comment here too? The user can always use a lock around e.g. with `concurrent::smart_lock`

Reply:
@mossmaurice 
The user can, but she cannot do as fine-grained as we can as she cannot access the internals (the update part). 

I would think it is a default use case for the following reason: I would use this `Servicediscovery` (or a custom one build on this) as a singleton, i.e. only one per application (either by using static or with some global). There is no point in creating multiple ones, since all they do is hog up memory. If I now have multiple threads I can easily use the ServiceDiscovery from each of them without worries as the `findService` call is protected by mutex.

It is different from e.g. subscribers which have a theoretically smaller scope and exist multiple times

Reply:
Ok, you convinced me, looking at the `PoshRuntime`:

```
bool PoshRuntimeImpl::sendRequestToRouDi(const IpcMessage& msg, IpcMessage& answer) noexcept
{
    // runtime must be thread safe
    std::lock_guard<posix::mutex> g(m_appIpcRequestMutex);
    return m_ipcChannelInterface.sendRequestToRouDi(msg, answer);
}
```

This means that all `getMiddlewareFoo()` methods are thread-safe.